### ISSUE TITLE
crash in systemd-resolve 248

### ISSUE THREAD

**Issue by mphilipps** [systemd/systemd#19931]
**systemd version the issue has been seen with**
 > 248, no issues with 247

**Used distribution**
 > Gentoo

**Linux kernel version used**
 > 5.4.97

**CPU architecture issue was seen on**
 > intel

**Expected behaviour you didn't see**
 > shouldn't crash

**Unexpected behaviour you saw**
 >  I updated systemd yesterday and today after booting I directly noticed multiple crashs of systemd-resolved.

```
[   36.874987] systemd-resolve[870]: segfault at 8 ip 00007fbcf9b09397 sp 00007ffe0d77dc90 error 4 in libsystemd-shared-248.so[7fbcf9afc000+193000]
[   36.875102] Code: 44 00 00 be 18 00 00 00 bf 01 00 00 00 e9 b1 59 ff ff 90 41 54 53 48 89 fb 48 83 ec 08 e8 31 c2 ff ff 49 89 c4 48 85 c0 74 2c <48> 8b 73 08 48 85 f6 75 30 48 8b 3b 48 c1 e6 03 e8 e4 c6 ff ff 49
```

Coredumpctl says the 

**Steps to reproduce the problem**
 > haven't been able to reproduce it yet, but it triggered multiple times during the first few minutes after boot.

**Additional program output to the terminal or log subsystem illustrating the issue**

```
/etc/systemd/resolved.conf
[Resolve]
DNS=8.8.8.8
DNSSEC=true
```

```
(gdb) bt
#0  bitmap_copy (b=0x0) at ../systemd-stable-248.3/src/shared/bitmap.c:38
#1  0x0000562dc6720189 in dns_resource_record_copy (rr=0x562dc827cef0) at ../systemd-stable-248.3/src/resolve/resolved-dns-rr.c:1645
#2  0x0000562dc67201fa in dns_resource_record_clamp_ttl (rr=rr@entry=0x7ffe0d77dd08, max_ttl=max_ttl@entry=3592)
    at ../systemd-stable-248.3/src/resolve/resolved-dns-rr.c:1705
#3  0x0000562dc66e44e2 in answer_add_clamp_ttl (answer=answer@entry=0x7ffe0d77dde0, rr=0x562dc827cef0, ifindex=5, answer_flags=67, rrsig=0x562dc827d1f0, 
    query_flags=query_flags@entry=131359, until=3629552183, current=36868845) at ../systemd-stable-248.3/src/resolve/resolved-dns-cache.c:952
#4  0x0000562dc66e606e in dns_cache_lookup (c=0x562dc82785b0, key=0x7ffe0d77dde0, query_flags=131359, ret_rcode=0x562dc8271fc0, 
    ret_answer=ret_answer@entry=0x562dc8271fb8, ret_full_packet=0x562dc8271fb0, ret_query_flags=0x562dc8271fd8, ret_dnssec_result=0x562dc8271fc4)
    at ../systemd-stable-248.3/src/resolve/resolved-dns-cache.c:1082
#5  0x0000562dc66fa33e in dns_transaction_prepare (t=0x562dc8271f80, ts=<optimized out>)
    at ../systemd-stable-248.3/src/resolve/resolved-dns-transaction.c:1707
#6  0x0000562dc66fa6d0 in dns_transaction_go (t=0x562dc8271f80) at ../systemd-stable-248.3/src/resolve/resolved-dns-transaction.c:1928
#7  0x0000562dc66e88be in dns_query_candidate_go (c=c@entry=0x562dc82c5330) at ../systemd-stable-248.3/src/resolve/resolved-dns-query.c:179
#8  0x0000562dc66e8d31 in dns_query_go (q=0x562dc826ed90) at ../systemd-stable-248.3/src/resolve/resolved-dns-query.c:808
#9  0x0000562dc66e8d95 in dns_query_go (q=<optimized out>) at ../systemd-stable-248.3/src/resolve/resolved-dns-query.c:823
#10 0x0000562dc66f3cc7 in dns_stub_process_query (m=0x562dc82629f0, l=0x0, s=0x0, p=0x562dc82a90c0)
    at ../systemd-stable-248.3/src/resolve/resolved-dns-stub.c:954
#11 0x0000562dc66f4039 in on_dns_stub_packet_internal (fd=<optimized out>, m=0x562dc82629f0, l=0x0, revents=<optimized out>, s=<optimized out>)
    at ../systemd-stable-248.3/src/resolve/resolved-dns-stub.c:976
#12 0x00007fbcf9c8a060 in source_dispatch (s=s@entry=0x562dc826f710) at ../systemd-stable-248.3/src/libsystemd/sd-event/sd-event.c:3512
#13 0x00007fbcf9c8a375 in dispatch_exit (e=<optimized out>) at ../systemd-stable-248.3/src/libsystemd/sd-event/sd-event.c:3660
#14 sd_event_dispatch (e=e@entry=0x562dc8265050) at ../systemd-stable-248.3/src/libsystemd/sd-event/sd-event.c:4070
#15 0x00007fbcf9c8aa48 in sd_event_run (e=e@entry=0x562dc8265050, timeout=timeout@entry=18446744073709551615)
    at ../systemd-stable-248.3/src/libsystemd/sd-event/sd-event.c:4138
#16 0x00007fbcf9c8ac6f in sd_event_loop (e=0x562dc8265050) at ../systemd-stable-248.3/src/libsystemd/sd-event/sd-event.c:4159
#17 0x0000562dc66dcc92 in run (argv=0x7ffe0d77e4c8, argc=1) at ../systemd-stable-248.3/src/resolve/resolved.c:92
#18 main (argc=1, argv=0x7ffe0d77e4c8) at ../systemd-stable-248.3/src/resolve/resolved.c:99
```

I am also using some link only dns setting for split view through a wireguard vpn.

**Comment by bluca** [comment]
If you can reproduce the crash reliably, would you be able to bisect it, given it works in v247? It would help quite a lot to speed up a fix.

**Comment by mphilipps** [comment]
I can't reproduce it yet, I made that ticket right in the morning and since then it has stopped crashing. Maybe it is boot related? I will check tomorrow again.

**Comment by poettering** [comment]
Almost certainly fixed by c856ef0457c35e9edfdbf085b69ec81c126d48e5.

Let#s close. If this is reproducible with git main we can reopen.
